crates/chia-consensus/src/gen/condition_tools.rs

```diff
@@ -0,0 +1,194 @@
+use crate::consensus_constants::ConsensusConstants;
+use crate::gen::opcodes::{
+    ConditionOpcode, AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT,
+    AGG_SIG_PARENT_PUZZLE, AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::OwnedSpend;
+use chia_protocol::Bytes;
+use chia_protocol::Coin;
+
+pub fn make_aggsig_final_message(
+    opcode: ConditionOpcode,
+    msg: &[u8],
+    spend: &OwnedSpend,
+    constants: &ConsensusConstants,
+) -> Vec<u8> {
+    let mut result = Vec::<u8>::with_capacity(msg.len() + 96);
+    result.extend(msg);
+    match opcode {
+        AGG_SIG_PARENT => {
+            result.extend(spend.parent_id.as_slice());
+            result.extend(constants.agg_sig_parent_additional_data.as_slice());
+        }
+        AGG_SIG_PUZZLE => {
+            result.extend(spend.puzzle_hash.as_slice());
+            result.extend(constants.agg_sig_puzzle_additional_data.as_slice());
+        }
+        AGG_SIG_AMOUNT => {
+            result.extend(u64_to_bytes(spend.coin_amount).as_slice());
+            result.extend(constants.agg_sig_amount_additional_data.as_slice());
+        }
+        AGG_SIG_PUZZLE_AMOUNT => {
+            result.extend(spend.puzzle_hash.as_slice());
+            result.extend(u64_to_bytes(spend.coin_amount).as_slice());
+            result.extend(constants.agg_sig_puzzle_amount_additional_data.as_slice());
+        }
+        AGG_SIG_PARENT_AMOUNT => {
+            result.extend(spend.parent_id.as_slice());
+            result.extend(u64_to_bytes(spend.coin_amount).as_slice());
+            result.extend(constants.agg_sig_parent_amount_additional_data.as_slice());
+        }
+        AGG_SIG_PARENT_PUZZLE => {
+            result.extend(spend.parent_id.as_slice());
+            result.extend(spend.puzzle_hash.as_slice());
+            result.extend(constants.agg_sig_parent_puzzle_additional_data.as_slice());
+        }
+        AGG_SIG_ME => {
+            let coin: Coin = Coin::new(spend.parent_id, spend.puzzle_hash, spend.coin_amount);
+
+            result.extend(coin.coin_id().as_slice());
+            result.extend(constants.agg_sig_me_additional_data.as_slice());
+        }
+        _ => return result,
+    };
+
+    result
+}
+
+pub(crate) fn u64_to_bytes(val: u64) -> Bytes {
```

Comment:
we already have two copies of this exact function (with this name) in the repository. The other copies are only available to tests however. So I still think it would make sense to make this private (not public) and use one of the existing copies for the test.

https://github.com/Chia-Network/chia_rs/pull/645